README.md
# koa-proxy [![Build Status](https://travis-ci.org/popomore/koa-proxy.png?branch=master)](https://travis-ci.org/popomore/koa-proxy) [![Coverage Status](https://coveralls.io/repos/popomore/koa-proxy/badge.png?branch=master)](https://coveralls.io/r/popomore/koa-proxy?branch=master)

Proxy middleware for koa

---

## Install

```
$ npm install koa-proxy -S
```

## Usage

When you request http://localhost:3000/index.js, it will fetch http://alicdn.com/index.js and return.

```js
var koa = require('koa');
var proxy = require('koa-proxy');
var app = koa();
app.use(proxy({
  host: 'http://alicdn.com'
}));
app.listen(3000);
```

You can proxy a specified url.

```js
app.get('index.js', proxy({
  url: 'http://alicdn.com/index.js'
}));
```

You can specify a key/value object that can map your request's path to the other.

```js
app.get('index.js', proxy({
  host: 'http://alicdn.com',
  map: {
    'index.js': 'index-1.js'
  }
}));
```

You can specify a function that can map your request's path to the desired destination.

```js
app.get('index.js', proxy({
  host: 'http://alicdn.com',
  map: function(path) { return 'public/' + path; }
}));
```

You can specify match criteria to restrict proxy calls to a given path.

```js
app.use(proxy({
  host:  'http://alicdn.com', // proxy alicdn.com...
  match: /^\/static\//        // ...just the /static folder
}));
```

Or you can use match to exclude a specific path.

```js
app.use(proxy({
  host:  'http://alicdn.com',     // proxy alicdn.com...
  match: /^(?!\/dontproxy\.html)/ // ...everything except /dontproxy.html
}));
```

Proxy won't send cookie to real server, you can set `jar = true` to send it.

```js
app.use(proxy({
  jar: true,
}));
```

Proxy won't send 'foo' and 'bar' headers to real server, or recieve 'jar-jar' from real server.

```js
app.use(proxy({
  suppressRequestHeaders: ['foo','bar'], // case-insensitive
  suppressResponseHeaders: ['jar-jar'] // case-insensitive
}));
```

You can use function to make host dependent on request

```js
app.use(proxy({
  host: (req) => req.header['x-host']
}));
```
## LICENSE

Copyright (c) 2014 popomore. Licensed under the MIT license.


HISTORY.md

0.9.0 / 2017-09-04
==================

  * Added suppressRequestHeaders and suppressResponseHeaders options (#54)

0.8.0 / 2017-02-03
==================

  * feat: [BREAKING_CHANGE] set jar = false by default (#48)
  * feat: add a yieldNext option (#47)

0.7.0 / 2016-12-27
==================

  * Add option to enable/disable following redirects (#45)
  * surely most people would want to install this as regular dependency, not as global one (#36)

0.6.0 / 2016-04-28
==================

  * Support requestOptions as functions (#28)

0.5.0 / 2016-02-24
==================

  * Exposing the cookie jar to the options interface
  * fix: should not return Transfer-encoding header
  * Added support for requestOptions to be passed as options to the underlying request
  * strip trailing slash from headers.host

0.4.1 / 2015-10-09
==================

  * check contentType is defined before calling indexOf (fixes bug when used in conjunction with koa-bodyparser)

0.4.0 / 2015-09-09
==================

  * added ability to have map as a function rather than an object

## 0.3.0

Add `host` and `match` option

## 0.2.0

Munge body object to JSON when appropriate

## 0.1.3

- fix headers option
- switch to co-request
- pass request method and body to proxy

## 0.1.2

return proxy's statusCode

## 0.1.1

should pass querystring

## 0.1.0

First commit


test/index.js
'use strict';

require('should');
var http = require('http');
var proxy = require('..');
var request = require('supertest');
var koa = require('koa');
var serve = require('koa-static');
var router = require('koa-router');
var parser = require('koa-body-parser');

describe('koa-proxy', function() {

  var server;
  before(function() {
    var app = koa();
    app.use(function* (next) {
      // Set this in response header to allow for proxy request header testing
      this.set('host', this.request.header.host);
      if (this.path === '/error') {
        this.body = '';
        this.status = 500;
        return;
      }
      if (this.path === '/postme') {
        this.body = this.req;
        this.set('content-type', this.request.header['content-type']);
        this.status = 200;
        return;
      }
      if (this.path === '/cookie-me') {
        this.cookies.set('test_cookie', 'nom-nom', { httpOnly: false });
        this.status = 200;
        return;
      }
      if (this.path === '/check-cookie') {
        this.body = this.cookies.get('test_cookie');
        this.status = 200;
        return;
      }
      if (this.path === '/redirect') {
        this.redirect('http://google.com');
        return;
      }
      if (this.path === '/suppress-my-headers') {
        var headers = this.request.header;
        this.set('jar-jar', 'binks');
        this.body = headers;
        this.status = 200;
        return;
      }
      if (this.querystring) {
        this.body = this.querystring;
        return;
      }
      yield* next;
    });
    app.use(serve(__dirname + '/fixtures'));
    server = app.listen(1234);
  });
  after(function() {
    server.close();
  });

  it('should have option url', function(done) {
    var app = koa();
    app.use(router(app));
    app.get('/index.js', proxy({
      url: 'http://localhost:1234/class.js'
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/index.js')
      .expect(200)
      .expect('Content-Type', /javascript/)
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.startWith('define("arale/class/1.0.0/class"');
        done();
      });
  });

  it('should have option url and host', function(done) {
    var app = koa();
    app.use(router(app));
    app.use(proxy({
      host: 'http://localhost:1234',
      url: 'class.js'
    }));
    app.get('/index.js', proxy({
      host: 'http://localhost:1234',
      url: 'class.js'
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/index.js')
      .expect(200)
      .expect('Content-Type', /javascript/)
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.startWith('define("arale/class/1.0.0/class"');
        done();
      });
  });

  it('should have option host', function(done) {
    var app = koa();
    app.use(proxy({
      host: 'http://localhost:1234'
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/class.js')
      .expect(200)
      .expect('Content-Type', /javascript/)
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.startWith('define("arale/class/1.0.0/class"');
        done();
      });
  });

  it('should strip trailing slash from option host', function(done) {
    var app = koa();
    app.use(proxy({
      host: 'http://localhost:1234/'
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/class.js')
      .expect(200)
      .expect('Host', 'localhost:1234')
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.startWith('define("arale/class/1.0.0/class"');
        done();
      });
  });

  it('should have option host and map', function(done) {
    var app = koa();
    app.use(proxy({
      host: 'http://localhost:1234',
      map: {
        '/index.js': '/class.js'
      }
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/index.js')
      .expect(200)
      .expect('Content-Type', /javascript/)
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.startWith('define("arale/class/1.0.0/class"');
        done();
      });
  });

  it('should have option host and map function', function(done) {
    var app = koa();
    app.use(proxy({
      host: 'http://localhost:1234',
      map: function(path) { return path.replace('index', 'class'); }
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/index.js')
      .expect(200)
      .expect('Content-Type', /javascript/)
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.startWith('define("arale/class/1.0.0/class"');
        done();
      });
  });

  it('should have option host and match', function(done) {
    var app = koa();
    app.use(proxy({
      host: 'http://localhost:1234',
      match: /^\/[a-z]+\.js$/
    }));
    app.use(proxy({
      host: 'http://localhost:1234'
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/class.js')
      .expect(200)
      .expect('Content-Type', /javascript/)
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.startWith('define("arale/class/1.0.0/class"');
        done();
      });
  });

  it('should have option followRedirect', function(done) {
    var app = koa();
    app.use(proxy({
      host: 'http://localhost:1234',
      followRedirect: false,
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/redirect')
      .expect(302)
      .expect('Location', /google.com/)
      .end(function (err, res) {
        if (err)
          return done(err);
        done();
      });
  });

  it('should have option yieldNext', function(done) {
    var app = koa();
    app.use(proxy({
      host: 'http://localhost:1234/',
      yieldNext: true,
    }));
    app.use(function* () {
      done();
    })
    var server = http.createServer(app.callback());
    request(server)
      .get('/class.js')
      .expect(200)
      .expect('Host', 'localhost:1234')
      .end(function (err, res) {
        if (err)
          return done(err);
      });
  });

  it('url not match for url', function(done) {
    var app = koa();
    app.use(proxy({
      url: 'class.js'
    }));
    app.use(function* () {
      this.body = 'next';
    });
    var server = http.createServer(app.callback());
    request(server)
      .get('/index.js')
      .expect(200)
      .expect('Content-Type', 'text/plain; charset=utf-8')
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.eql('next');
        done();
      });
  });

  it('url not match for map', function(done) {
    var app = koa();
    app.use(proxy({
      map: {
        '/index.js': '/class.js'
      }
    }));
    app.use(function* () {
      this.body = 'next';
    });
    var server = http.createServer(app.callback());
    request(server)
      .get('/index.js')
      .expect(200)
      .expect('Content-Type', 'text/plain; charset=utf-8')
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.eql('next');
        done();
      });
  });

  it('option exist', function() {
    (function() {
      proxy();
    }).should.throw();
  });

  it('encoding', function(done) {
    var app = koa();
    app.use(proxy({
      url: 'http://localhost:1234/index.html',
      encoding: 'gbk'
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/index.js')
      .expect(200)
      .expect('Content-Type', 'text/html; charset=utf-8')
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.startWith('<div>中国</div>');
        done();
      });
  });

  it('pass query', function(done) {
    var app = koa();
    app.use(proxy({
      url: 'http://localhost:1234/class.js',
      encoding: 'gbk'
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/index.js?a=1')
      .expect(200)
      .expect('Content-Type', 'text/plain; charset=utf-8')
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.startWith('a=1');
        done();
      });
  });

  it('pass request body', function(done) {
    var app = koa();
    app.use(proxy({
      host: 'http://localhost:1234',
    }));
    var server = http.createServer(app.callback());
    request(server)
      .post('/postme')
      .send({foo:'bar'})
      .expect(200)
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.equal('{"foo":"bar"}');
        done();
      });
  });

  it('pass parsed request body', function(done) {
    var app = koa();
    app.use(parser()); // sets this.request.body
    app.use(proxy({
      host: 'http://localhost:1234',
    }));
    var server = http.createServer(app.callback());
    request(server)
      .post('/postme')
      .send({foo:'bar'})
      .expect(200)
      .end(function (err, res) {
        if (err)
          return done(err);
        res.text.should.equal('{"foo":"bar"}');
        done();
      });
  });

  it('statusCode', function(done) {
    var app = koa();
    app.use(proxy({
      host: 'http://localhost:1234'
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/error')
      .expect(500, done);
  });

  it('should pass along requestOptions', function(done) {
    var app = koa();
    app.use(proxy({
      url: 'http://localhost:1234/class.js',
      requestOptions: { timeout: 1 }
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/index.js')
      .expect(function sleep() {
        // Using the custom assert function to make sure we get a timeout
        var sleepTime = new Date().getTime() + 3;
        while(new Date().getTime() < sleepTime) {}
      })
      .expect(500, done);
  });

  it('should pass along requestOptions when function', function(done) {
    var app = koa();
    app.use(proxy({
      url: 'http://localhost:1234/class.js',
      requestOptions: function(req, opt) {
        opt.timeout = 1;
        return opt;
      }
    }));
    var server = http.createServer(app.callback());
    request(server)
      .get('/index.js')
      .expect(function sleep() {
        // Using the custom assert function to make sure we get a timeout
        var sleepTime = new Date().getTime() + 3;
        while(new Date().getTime() < sleepTime) {}
      })
      .expect(500, done);
  });

  describe('with cookie jar', function () {

    var app = koa();
    app.use(router(app));
    app.use(proxy({
      host: 'http://localhost:1234',
      jar: true
    }));
    var server = http.createServer(app.callback());
    var agent = request.agent(server);

    it('should set cookies', function(done) {
      agent
        .get('/cookie-me')
        .expect(200)
        .expect('set-cookie', 'test_cookie=nom-nom; path=/')
        .end(function (err, res) {
          if (err)
            return done(err);
          done();
        });
    });

    it('should retain cookies', function(done) {
      agent
        .get('/check-cookie')
        .expect(200)
        .expect('nom-nom', done);
    });

    it('should retain cleared cookies', function(done) {
      var req = agent
        .get('/check-cookie');
      req.cookies = '';
      req
        .expect(200)
        .expect('nom-nom', done);
    });

  });

  describe('without cookie jar', function () {

    var app = koa();
    app.use(router(app));
    app.use(proxy({
      host: 'http://localhost:1234',
    }));
    var server = http.createServer(app.callback());
    var agent = request.agent(server);

    it('should set cookies', function(done) {
      agent
        .get('/cookie-me')
        .expect(200)
        .expect('set-cookie', 'test_cookie=nom-nom; path=/')
        .end(function (err, res) {
          if (err)
            return done(err);
          done();
        });
    });

    it('should retain cookies', function(done) {
      agent
        .get('/check-cookie')
        .expect(200)
        .expect('nom-nom', done);
    });

    it('should not retain cleared cookies', function(done) {
      var req = agent
        .get('/check-cookie');
      req.cookies = '';
      req
        .expect(200)
        .end(function (err, res) {
          if (err)
            return done(err);
          res.text.should.not.equal('nom-nom');
          done();
        });
    });
  });

  describe('with suppressed request and response headers', function () {

    var app = koa();
    app.use(router(app));
    app.use(proxy({
      host: 'http://localhost:1234',
      suppressRequestHeaders: ['foO','bAr'],
      suppressResponseHeaders: ['jaR-Jar']
    }));
    var server = http.createServer(app.callback());
    var agent = request.agent(server);

    it('should remove headers', function(done) {
      agent
        .get('/suppress-my-headers')
        .set('foo', 'kung')
        .set('bar', 'none')
        .set('happy', 'path')
        .expect(200)
        .expect(function(res){
          if(res.headers['jar-jar']) throw new Error("jar-jar header was present");
          if(res.body['foo']) throw new Error("foo header was present");
          if(res.body['bar']) throw new Error("bar header was present");
        })
        .end(function (err, res) {
          if (err)
            return done(err);
          done();
        });
    });
  });
});


index.js
'use strict';

var join = require('url').resolve;
var iconv = require('iconv-lite');
var coRequest = require('co-request');

module.exports = function(options) {
  options || (options = {});
  var request = coRequest.defaults({ jar: options.jar === true });

  if (!(options.host || options.map || options.url)) {
    throw new Error('miss options');
  }

  return function* proxy(next) {
    var host = options.host;
    if (typeof options.host === 'function') {
      host = options.host(this.request);
    }

    var url = resolve(host, this.path, options);

    if(typeof options.suppressRequestHeaders === 'object'){
      options.suppressRequestHeaders.forEach(function(h, i){
        options.suppressRequestHeaders[i] = h.toLowerCase();
      });
    }

    var suppressResponseHeaders = [];  // We should not be overwriting the options object!
    if(typeof options.suppressResponseHeaders === 'object'){
      options.suppressResponseHeaders.forEach(function(h, i){
        suppressResponseHeaders.push(h.toLowerCase());
      });
    }

    // don't match
    if (!url) {
      return yield* next;
    }

    // if match option supplied, restrict proxy to that match
    if (options.match) {
      if (!this.path.match(options.match)) {
        return yield* next;
      }
    }

    var parsedBody = getParsedBody(this);

    var opt = {
      url: url + (this.querystring ? '?' + this.querystring : ''),
      headers: this.header,
      encoding: null,
      followRedirect: options.followRedirect === false ? false : true,
      method: this.method,
      body: parsedBody,
    };

    // set 'Host' header to host (without protocol prefix), strip trailing slash
    if (host) opt.headers.host = host.slice(host.indexOf('://')+3).replace(/\/$/,'');

    if (options.requestOptions) {
      if (typeof options.requestOptions === 'function') {
        opt = options.requestOptions(this.request, opt);
      } else {
        Object.keys(options.requestOptions).forEach(function (option) { opt[option] = options.requestOptions[option]; });
      }
    }

    for(name in opt.headers){
      if(options.suppressRequestHeaders && options.suppressRequestHeaders.indexOf(name.toLowerCase()) >= 0){
        delete opt.headers[name];
      }
    }

    var requestThunk = request(opt);

    if (parsedBody) {
      var res = yield requestThunk;
    } else {
      // Is there a better way?
      // https://github.com/leukhin/co-request/issues/11
      var res = yield pipeRequest(this.req, requestThunk);
    }

    this.status = res.statusCode;
    for (var name in res.headers) {
      // http://stackoverflow.com/questions/35525715/http-get-parse-error-code-hpe-unexpected-content-length
      if(suppressResponseHeaders.indexOf(name.toLowerCase())>=0){
        continue;
      }
      if (name === 'transfer-encoding') {
        continue;
      }
      this.set(name, res.headers[name]);
    }

    if (options.encoding === 'gbk') {
      this.body = iconv.decode(res.body, 'gbk');
      return;
    }

    this.body = res.body;

    if (options.yieldNext) {
      yield next;
    }
  };
};


function resolve(host, path, options) {
  var url = options.url;
  if (url) {
    if (!/^http/.test(url)) {
      url = host ? join(host, url) : null;
    }
    return ignoreQuery(url);
  }

  if (typeof options.map === 'object') {
    if (options.map && options.map[path]) {
      path = ignoreQuery(options.map[path]);
    }
  } else if (typeof options.map === 'function') {
    path = options.map(path);
  }

  return host ? join(host, path) : null;
}

function ignoreQuery(url) {
  return url ? url.split('?')[0] : null;
}

function getParsedBody(ctx){
  var body = ctx.request.body;
  if (body === undefined || body === null){
    return undefined;
  }
  var contentType = ctx.request.header['content-type'];
  if (!Buffer.isBuffer(body) && typeof body !== 'string'){
    if (contentType && contentType.indexOf('json') !== -1){
      body = JSON.stringify(body);
    } else {
      body = body + '';
    }
  }
  return body;
}

function pipeRequest(readable, requestThunk){
  return function(cb){
    readable.pipe(requestThunk(cb));
  }
}


index2.js
'use strict';

const join = require('url').resolve;
const iconv = require('iconv-lite');
const asyncRequest = require('./libs/async-request');

module.exports = function(options) {
	options || (options = {})
	var request = asyncRequest.defaults({
		cookieJar: options.jar === true,
	});
	if (!(options.host || options.map || options.url)) {
		throw new Error('miss options');
	}
	return async function(ctx, next) {
		let url = resolve(ctx.path, options);
		if (!url) {
			return await next();
		}
		// if match option supplied, restrict proxy to that match
		if (options.match) {
			if (!url.match(options.match)) {
				return await next();
			}
		}
		let parsedBody = getParsedBody(ctx);
		url = url + (ctx.request.querystring ? '?' + ctx.request.querystring : '');
		var opt = {
			uri: url,
			headers: ctx.request.headers,
			encoding: null,
			followRedirect: options.followRedirect === false ? false : true,
			method: ctx.request.method,
			body: parsedBody,
		};

		// set 'Host' header to options.host (without protocol prefix), strip trailing slash
		if (options.host) opt.headers.host = options.host.slice(options.host.indexOf('://') + 3).replace(/\/$/, '');
		if (options.requestOptions) {
			if (typeof options.requestOptions === 'function') {
				opt = options.requestOptions(ctx.request, opt);
			} else {
				Object.keys(options.requestOptions).forEach(function(option) {
					opt[option] = options.requestOptions[option];
				});
			}
		}
		/*async request*/
		try {
			var res = await request(url, opt);
		} catch (e) {
			//some handle
			// throw new Error("request error" + e);
			console.log("catch error" + e);
		}
		ctx.response.status = res ? res.statusCode : 500;
		//set the response fields
		let responseHeaders = res ? res.headers : null;
		let responseBody = res ? res.body : null;
		ctx.response.set(responseHeaders);
		if (options.encoding === 'gbk') {
			ctx.response.body = iconv.decode(res.body, 'gbk');
			return;
		}
		ctx.response.body = responseBody;
		/*maybe should change this opts*/
		if (options.yieldNext) {
			await next();
		}
	}
}

/*util method*/
function getParsedBody(ctx) {
	var body = ctx.request.body;
	if (body === undefined || body === null) {
		return undefined;
	}
	var contentType = ctx.request.header['content-type'];
	if (!Buffer.isBuffer(body) && typeof body !== 'string') {
		if (contentType && contentType.indexOf('json') !== -1) {
			body = JSON.stringify(body);
		} else {
			body = body + '';
		}
	}
	return body;
}

function resolve(path, options, local) {
	var url = options.url;
	if (url) {
		if (!/^http/.test(url)) {
			url = options.host ? join(options.host, url) : null;
			if (options.port) {
				//check agagin(硬编码强行支持指向本机文件,暂不知道v1版本为什么不会有这个问题)
				url = "http://localhost:" + options.port + "/" + options.url;
			}
		}
		return ignoreQuery(url);
	}

	if (typeof options.map === 'object') {
		if (options.map && options.map[path]) {
			path = ignoreQuery(options.map[path]);
		}
	} else if (typeof options.map === 'function') {
		path = options.map(path);
	}
	return options.host ? join(options.host, path) : null;
}

function ignoreQuery(url) {
	return url ? url.split('?')[0] : null;
}

libs/async-request.js
'use strict';


/* fork https://github.com/gajus/async-request*/
let Arequest,
    request = require('request'),
    _ = require('lodash');

request.debug = false;

Arequest = (defaultOptions) => {
    let arequest;

    Arequest.validateOptions(defaultOptions);

    arequest = async(url, options) => {
        return new Promise((resolve) => {
            Arequest.validateOptions(options);

            options = _.assign(options, defaultOptions);

            options = Arequest.mapOptions(options);
            request(url, options, (error, response) => {
                if (error) {
                    throw new Error(error);
                }

                resolve({
                    statusCode: response.statusCode,
                    headers: response.headers,
                    body: response.body
                });
            });
        });
    };

    arequest.defaults = (options) => {
        if (!options) {
            return defaultOptions;
        }

        if (options.cookieJar === true) {
            options.cookieJar = request.jar();
        }

        return Arequest(options);
    };

    return arequest;
};

/**
 *
 */
Arequest.validateOptions = (options) => {

    if (!options) {
        return;
    }
    if (options.method && _.indexOf(['GET', 'POST', 'PUT', 'HEAD', 'DELETE'], options.method) === -1) {
        throw new Error('Unknown option.method value ("' + options.method + '").');
    }
};

/**
 * Map options to meet the request interface.
 */
Arequest.mapOptions = (options) => {
    if (!options) {
        return options;
    }

    if (options.data) {
        options.form = options.data;

        delete options.data;
    }

    if (options.cookieJar) {
        options.jar = options.cookieJar;

        delete options.cookieJar;
    }

    return options;
};

module.exports = Arequest({});

app.js
const koa = require('koa');
const app = new koa();
const proxy = require('./index2');
var Router = require('koa-router');
var serve = require('koa-static');
var parser = require('koa-bodyparser');
var router = new Router();
app.use(parser())
app.use(proxy({
	host: 'http://localhost:3008',
}));
app.use(async(ctx, next) => {
	// Set this in response header to allow for proxy request header testing
	ctx.set('host', ctx.request.header.host);
	if (ctx.path === '/error') {
		ctx.response.body = '';
		ctx.response.status = 500;
		return;
	}
	if (ctx.path === '/postme') {
		console.log("??")
		ctx.response.body = ctx.request.body;
		ctx.set('content-type', ctx.request.header['content-type']);
		ctx.response.status = 200;
		return;
	}
	if (ctx.path === '/cookie-me') {
		ctx.cookies.set('test_cookie', 'nom-nom', {
			httpOnly: false
		});
		ctx.response.status = 200;
		return;
	}
	if (ctx.path === '/check-cookie') {
		ctx.response.body = ctx.cookies.get('test_cookie');
		ctx.response.status = 200;
		return;
	}
	if (ctx.path === '/redirect') {
		ctx.redirect('http://google.com');
		return;
	}

	if (ctx.querystring) {
		ctx.response.body = ctx.querystring;
		return;
	}
	await next();
});
app.use(serve(__dirname + '/test/fixtures'));
// router.get('/index.js', proxy({
// 	url: 'http://localhost:3008/class.js'
// }));
// app.use(proxy({
// 	url: 'class.js',
// 	port: "3008"
// }));

app
	.use(router.routes())


app.listen(3008);